invenio_rdm_records/records/systemfields/tombstone.py

```diff
@@ -0,0 +1,157 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 TU Wien.
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+
+from datetime import datetime
+
+from invenio_records.systemfields import SystemField
+from invenio_requests.resolvers.registry import ResolverRegistry
+
+
+class Tombstone:
+    """A data class for managing tombstone information."""
+
+    def __init__(self, data):
+        """Constructor."""
+        self.removal_reason = data.get("removal_reason")
+        self.note = data.get("note")
+        self.removed_by = data.get("removed_by")
+        self.removal_date = data.get("removal_date")
+
+    @property
+    def removal_reason(self):
+        """Get the removal reason."""
+        return self._removal_reason
+
+    @removal_reason.setter
+    def removal_reason(self, value):
+        """Set the removal reason, referencing the corresponding vocabulary.
+
+        If the value is a string, it will be turned into a dictionary of the
+        shape ``{"id": value}``.
+        """
+        # NOTE: the validity of the reference is checked
+        #       in the record's relations systemfield
+        if isinstance(value, str):
+            value = {"id": value}
+        self._removal_reason = value or None
+
+    @property
+    def note(self):
+        """Get the public note attached to the tombstone."""
+        return self._note
+
+    @note.setter
+    def note(self, value):
+        """Set the public note attached to the tombstone."""
+        self._note = value or ""
+
+    @property
+    def removed_by(self):
+        """Get a reference to the entity who removed the record."""
+        return self._removed_by
+
+    @removed_by.setter
+    def removed_by(self, value):
+        """Set a reference to the entity who removed the record.
+
+        If the value is a string or integer, it is assumed that the
+        reference is a user and the value will be turned into a dictionary
+        of the shape ``{"user": value}``.
+        If the value is a referenceable entity (such as a ``User`` object),
+        the ``ResolverRegistry`` will be used to create a reference dict.
+        """
+        if value is None:
+            self._removed_by = None
+
+        elif (ref := ResolverRegistry.reference_entity(value)) is not None:
```

Comment:
that's cool but if it's just that statement that forces us to move from 3.7 maybe is too aggressive now? ping @lnielsen @ntarocco

Reply:
just as a reminder, python 3.7 has reached end of life already: https://devguide.python.org/versions/

Reply:
yes, we also started removing from tests but we were discussing not forcing people to upgrade their python version...so my comment was related to this topic...

Reply:
As a general policy for Invenio, we want to try keeping the window open to allow older versions of libraries/runtimes running without completely failing (i.e. not "forcing" version upgrades). That's also why we still keep `python_requires = >=3.7` in `setup.cfg`. If other dependencies though start forcing us to drop v3.7, that's our cue to also officially do it, since the effort will become too much to maintain this "implicit" backwards compat.

I agree that in this case the walrus operator probably increases readability (vs the alternative of adding an extra level of nesting inside the `else:`). 

To play the devil's advocate though:

- if we're not testing this in CI, the only way to find out that Python 3.7 is broken is to wait for someone to report it...
- I see that e.g. even newer versions of Flask have [added `python_requires = >=3.8`](https://github.com/pallets/flask/blob/b1385919bea3238d49a6e2cfc8261771775cd6e9/pyproject.toml#L20). So we're currently in this limbo state where because we've upper-pinned Flask v2.3, we get a "free pass" for Python v3.7... I assume that once a security fix comes in Flask v2.3.x, and not v2.2.x, we must buckle up and make the move (and then be actually forced to drop Python 3.7).

At this point, I'm pro towards dropping Python v3.7 support and using Python v3.8+ features since if the 3.7 build is not failing already, it will happen anyways once we unpin Flask.

@ntarocco WDYT?

Reply:
Discussed with Nico IRL, and it's good to go as long as we're not mentioning anywhere in the docs explicitly that we're supporting Python v3.7 still.